### PR TITLE
DDF-3197 Disable the felix file installer and manually sync etc with config admin

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Disables the felix file installer, which writes config admin changes back to `etc`, so we can enable our own installer that correctly supports writing **and** deleting. 

**NOTE**: Unit test changes are pending first round of feedback. 

#### Who is reviewing it? 
@paouelle 
@brjeter 
@oconnormi 
@dannythk 
@vinamartin 

#### Choose 2 committers to review/merge the PR. 
@coyotesqrl
@figliold

#### How should this be tested? (List steps with links to updated documentation)
Start up DDF with defaults for everything. 
Keep the logs tailed throughout this process to catch any infinite config loops. 

1. Setup various configs, both singleton-type and factory-type configurations. 
✅  Verify a config in `etc` gets created for each config created using the UI

2. Delete one or two configs using the UI
✅  Verify their corresponding files in `etc` get deleted as well

3. Delete one or two configs in `etc` and refresh the admin UI
✅  Verify the configs no longer show up in the admin UI

4. Shutdown the system, delete one or two configs in `etc`, and reboot the system
✅  Verify the configs no longer show up in the admin UI

#### Any background context you want to provide?
Highly available systems that are configured using the UI before being switched to a proper solution such as Puppet will need this kind of coherency between `etc` and the bundle cache. Also, this will be a necessary plugin point for encrypting passwords found in the `etc` directory. 

##### This ticket:
[DDF-3197](https://codice.atlassian.net/browse/DDF-3197)
##### To be done after:
[DDF-3185](https://codice.atlassian.net/browse/DDF-3185)
##### Prereq's already done:
[DDF-3186](https://codice.atlassian.net/browse/DDF-3186)
[DDF-3148](https://codice.atlassian.net/browse/DDF-3148)

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java [hunk 1]
```diff
@@ -135,31 +171,38 @@ private void handleUpdate(ConfigurationEvent event) {
...
 
-        // Config not tracked and will not be tracked? (1)
-        if (felixFileOriginal == null && felixFileNew == null) {
-            LOGGER.debug("Was not tracked and will not track pid {}", pid);
+        if (felixFileOriginal == null) {
+            // Config not tracked? (1)
+            if (felixFileNew == null) {
+                felixFileNew = felixConfig.generateDefaultFelixFile()
+                        .getFelixFile();
+            }
```
Comment (Lambeaux) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: I'll be honest, I don't like what I'm doing here and I'm open to suggestions. I thought: 
- Make `FelixConfig` immutable, but that doesn't make sense because it wraps a `Configuration` and _will_ mutate it if instructed to do so. 
- Stop using local variables and just use `felixConfig.getFelixFile()` everywhere but that strikes me as verbose.

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java [hunk 2]
```diff
@@ -135,31 +171,38 @@ private void handleUpdate(ConfigurationEvent event) {
...
+        }
+
+        // Respond to config properties being updated (4)
+        // Updating properties on disk, if they haven't changed, causes an infinite loop
+        // with the felix DirectoryWatcher
+        Path path = felixFileOriginal.toPath();
+        Dictionary<String, Object> fileState = configActions.getProperties(path);
+        Dictionary<String, Object> configState = felixConfig.getSanitizedProperties();
+        if (!equalDictionaries(configState, fileState)) {
+            performOperation(configActions.update(path, configState, false));
```
Comment (Lambeaux) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: **TODO**: Cache `fileState` as part of the map in this object so we don't have to read from disk every time. Do we agree?

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java [hunk 3]
```diff
@@ -69,29 +86,46 @@ File getFelixFile() {
...
                 file.toURI()
                         .toString());
-        config.update(propsWithFelixFileName);
+        try {
+            config.update(propsWithFelixFileName);
+            this.felixFile = file;
+        } catch (IOException e) {
+            LOGGER.error("Error writing configuration");
+            LOGGER.debug("Could not set felix file name, error writing configuration: ", e);
+            this.felixFile = null;
```
Comment (Lambeaux) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: It might be time to make the `felixFile` field be an `Optional`, thoughts?

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java [hunk 4]
```diff
@@ -69,29 +86,46 @@ File getFelixFile() {
...
+    /**
+     * Sets the felix file name using the default technique from the config repo, using .config by
+     * default in case complex data structures are present.
+     *
+     * @return <code>this</code> for chaining.
+     */
+    FelixConfig generateDefaultFelixFile() {
+        setFelixFile(createFelixFileName(config));
+        return this;
+    }
```
Comment (Lambeaux) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: I'm not sure what to do with this sort of "default setter". I'm open to ideas.

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 5]
```diff
@@ -34,58 +45,244 @@
  */
 public abstract class ConfigOperation implements Operation<Void> {
```
Comment (Lambeaux) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: @coyotesqrl let me know if there's anything I can do to improve upon these impls. They are very similar to their property counterparts but they operate on Dictionaries.

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 6]
```diff
@@ -34,58 +45,244 @@
...
+            } catch (IOException e) {
+                return ResultImpl.fail(new ConfiguratorException("Error creating config: ", e));
+            }
+        }
+
+        /**
+         * If a config was created on top of an already existing file, a successful rollback would
+         * cause the file to be deleted, which was NOT precisely the state the system was in before.
+         * The caller is responsible for invoking the appropriate operation knowing full well the
+         * rollback consequences.
```
Comment (Lambeaux) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: Is this the right way to reason about rollback?

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java (hunk 2 above)
Comment (paouelle) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: We definitely do not want to reload from disk every time. Do you think caching the whole map would increase memory requirement too much? If so maybe just caching a checksum but then we have to compute that checksum every time. So I guess it is memory vs processing. But since config objects should not change that often, maybe the checksum alone is good enough.

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java (hunk 3 above)
Comment (paouelle) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: Typically we leave optional on the interface level and not inside as attribute's value. But I have never been strict on that one ;-)

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 7]
```diff
@@ -34,58 +45,244 @@
...
+                        "Problem deleting properties file " + configFile));
             }
             return ResultImpl.pass();
         }
 
         @Override
         public Result<Void> rollback() throws ConfiguratorException {
-            return ResultImpl.rollback();
+            try {
+                createConfig(configFile, strategy, originalConfigs);
```
Comment (coyotesqrl) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: Just like you attempt to restore the original file here on rollback, that's what you should do in the case of an overwrite.

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 8]
```diff
@@ -34,58 +45,244 @@
...
+            Dictionary<String, Object> configsToWrite = new Hashtable<>();
+            originalConfigs = readConfig(configFile, strategy);
+
+            if (keepIfNotPresent) {
+                putAll(originalConfigs, configsToWrite);
+            }
+            putAll(configs, configsToWrite);
+
+            try {
+                FileOutputStream outputStream = new FileOutputStream(configFile);
```
Comment (coyotesqrl) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: ❗️ Try-with-resources so the output stream closes.

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 9]
```diff
@@ -34,58 +45,244 @@
...
+                return ResultImpl.pass();
+            } catch (IOException e) {
+                return ResultImpl.fail(new ConfiguratorException("Error writing config: ", e));
+            }
+        }
+
+        @Override
+        public Result<Void> rollback() throws ConfiguratorException {
+            try {
+                FileOutputStream outputStream = new FileOutputStream(configFile);
```
Comment (coyotesqrl) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: ❗️ Try-with-resources so the output stream closes.

**Comment by dannythk** [comment]
Since I am not a coder, I blackbox verified the functionality.

Test with the following configurations:
1 - ddf.services.schematron.SchematronValidationService-SchematronValidators.ddms20.validation.cfg and ddf.services.schematron.SchematronValidationService-SchematronValidators.ddms50.ism.cfg
2 - ddf.platform.ui.config
3 - org.codice.ddf.security.policy.context.impl.PolicyManager

Verified the following scenarios
1 - Added configuration via admin UI and verified the corresponding config file is created in /etc folder
2 - Removed configuration via admin UI and verified the corresponding config file is removed in /etc folder
3 - Added configuration file in the /etc folder, verified the configuration showed up on the admin UI (live DDF)
4 - Remove configuration file in the /etc folder, verified the configuration is removed from the admin UI (live DDF)
5 - Added configuration file in the /etc/ folder, restart DDF, verified the configuration showed up on the admin UI 
6 - Remove configuration file in the /etc folder, restart DDF, verified the configuration is removed from the admin UI
7 - Tried to put in an invalid configuration file with the proper file name, DDF does split out error message stating it failed to initialized the component.

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java [hunk 10]
```diff
@@ -171,7 +214,7 @@ private void handleUpdate(ConfigurationEvent event) {
     private void handleDelete(ConfigurationEvent event) {
         String pid = event.getPid();
         File felixFileOriginal = pidFileMap.remove(pid);
```
Comment (figliold) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: ❓ Is it possible that the map and file system get out of sync if performOperation() fails?

File: platform/admin/configurator/admin-configurator-actions-api/src/main/java/org/codice/ddf/internal/admin/configurator/actions/ConfigActions.java [hunk 11]
```diff
@@ -13,10 +13,59 @@
...
+     * @param configFile       the config file to be updated
+     * @param configs          dictionary of key:value pairs to be written to the config file
+     * @param keepIfNotPresent if true, any keys in the current config file that are not in the
+     *                         {@code configs} map will be left with their initial values; if false, they
+     *                         will be removed from the file
+     * @return instance of this class
+     * @throws ConfiguratorException if an error occurs creating the operator
+     */
+    Operation<Void> update(Path configFile, Dictionary<String, Object> configs,
+            boolean keepIfNotPresent) throws ConfiguratorException;
```
Comment (brjeter) [inline platform/admin/configurator/admin-configurator-actions-api/src/main/java/org/codice/ddf/internal/admin/configurator/actions/ConfigActions.java]: ✏️ Should we even keep this `keepIfNotPresent` property around if we're only using it for the case that it's false? We could just document that the behavior is all additional properties will be removed.

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java [hunk 12]
```diff
@@ -207,4 +251,17 @@ private Configuration getConfiguration(ConfigurationEvent configurationEvent) {
             return null;
         }
     }
+
+    private static boolean equalDictionaries(Dictionary x, Dictionary y) {
+        if (x.size() != y.size()) {
+            return false;
+        }
+        for (final Enumeration e = x.keys(); e.hasMoreElements(); ) {
```
Comment (figliold) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: :question: raw type for Enumeration?

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java [hunk 13]
```diff
@@ -69,29 +86,46 @@ File getFelixFile() {
...
 
-        this.felixFile = file;
+    /**
+     * Deletes the configuration from config admin.
+     */
+    void delete() {
+        try {
+            config.delete();
+        } catch (IOException e) {
+            LOGGER.error("Error deleting configuration");
```
Comment (figliold) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: :question: Is it okay to just log if the delete fails?

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java [hunk 14]
```diff
@@ -69,29 +86,46 @@ File getFelixFile() {
...
         Dictionary<String, Object> propsWithFelixFileName = config.getProperties();
         propsWithFelixFileName.put(FELIX_FILENAME_PROP,
                 file.toURI()
                         .toString());
-        config.update(propsWithFelixFileName);
+        try {
+            config.update(propsWithFelixFileName);
+            this.felixFile = file;
+        } catch (IOException e) {
+            LOGGER.error("Error writing configuration");
```
Comment (figliold) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: :question: is it okay to just log if the update fails?

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 15]
```diff
@@ -34,58 +45,244 @@
...
+            this.originalConfigs = new Hashtable<>();
+        }
+
+        @Override
+        public Result<Void> commit() throws ConfiguratorException {
+            Dictionary<String, Object> configsToWrite = new Hashtable<>();
+            originalConfigs = readConfig(configFile, strategy);
+
+            if (keepIfNotPresent) {
+                putAll(originalConfigs, configsToWrite);
```
Comment (brjeter) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: ❓ Is there a reason we didn't do `configsToWrite.putAll(originalConfigs)`?

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 16]
```diff
@@ -34,58 +45,244 @@
...
+            }
+            putAll(configs, configsToWrite);
```
Comment (brjeter) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: ❓ Same question as above about `putAll`

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 17]
```diff
@@ -34,58 +45,244 @@
...
+        }
+
+        @Override
+        public String toString() {
+            return "Updating config " + configFile.getAbsolutePath();
+        }
+
+        private static void putAll(Dictionary<String, Object> source,
+                Dictionary<String, Object> target) {
+            Enumeration keys = source.keys();
```
Comment (figliold) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: :question: raw type for Enumeration?

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java [hunk 18]
```diff
@@ -34,58 +45,244 @@
...
+            while (keys.hasMoreElements()) {
+                String key = (String) keys.nextElement();
```
Comment (figliold) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: If you use Enumeration<String> keys = ... above, you shouldn't need the cast here.

File: platform/admin/configurator/admin-configurator-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 19]
```diff
@@ -14,6 +14,14 @@
 -->
 <blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0">
 
+    <bean id="strategiesBean" class="org.codice.ddf.platform.util.SortedServiceList"/>
+    <reference-list id="strategiesRefList"
+                    interface="org.codice.ddf.platform.io.internal.PersistenceStrategy"
+                    availability="optional">
+        <reference-listener ref="strategiesBean" bind-method="bindPlugin"
```
Comment (figliold) [inline platform/admin/configurator/admin-configurator-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: :question: Do you have bindPlugin and unbindPlugin methods?

File: platform/admin/configurator/admin-configurator-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 20]
```diff
@@ -14,6 +14,14 @@
 -->
 <blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0">
 
+    <bean id="strategiesBean" class="org.codice.ddf.platform.util.SortedServiceList"/>
+    <reference-list id="strategiesRefList"
+                    interface="org.codice.ddf.platform.io.internal.PersistenceStrategy"
```
Comment (figliold) [inline platform/admin/configurator/admin-configurator-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Can this really be optional?

File: platform/platform-app/src/main/resources/features.xml [hunk 21]
```diff
@@ -1030,6 +1030,12 @@
         <bundle>mvn:ddf.platform.error/platform-error-impl/${project.version}</bundle>
     </feature>
 
+    <feature name="io" install="manual" version="${project.version}"
+             description="Internal tools for general purpose IO functions">
```
Comment (figliold) [inline platform/platform-app/src/main/resources/features.xml]: :question: are these really general tools for IO functions or are they really specific to reading and writing config admin config files?

File: platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java (hunk 5 above)
Comment (paouelle) [inline platform/admin/configurator/admin-configurator-impl/src/main/java/org/codice/ddf/admin/configurator/impl/ConfigOperation.java]: ❓ Where are the unit tests?

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java [hunk 22]
```diff
@@ -18,13 +18,19 @@
...
+import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
+import javax.annotation.Nullable;
+
 import org.codice.ddf.admin.configurator.Configurator;
 import org.codice.ddf.admin.configurator.ConfiguratorException;
 import org.codice.ddf.admin.configurator.ConfiguratorFactory;
```
Comment (paouelle) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: ✏️  I think we need to add more test cases for ConfigurationInstaller.

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java [hunk 23]
```diff
@@ -18,13 +18,19 @@
```
Comment (paouelle) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java]: ❗️ Are we guaranteed that ConfigAdmin will not call this listener multiple times at the same time from different threads? If not then we should be careful about synchronization as we check and update internal attributes in the same method at different places.

File: platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java [hunk 24]
```diff
@@ -69,29 +86,46 @@ File getFelixFile() {
...
-    /**
      * Sets the felix file name to the provided {@link File} argument.
      *
      * @param file the file to use for the felix file name property.
-     * @throws IOException if an error occurs persisting to config admin.
      */
-    void setFelixFile(File file) throws IOException {
+    void setFelixFile(File file) {
         Dictionary<String, Object> propsWithFelixFileName = config.getProperties();
         propsWithFelixFileName.put(FELIX_FILENAME_PROP,
```
Comment (paouelle) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: @brjeter I think this method's intention was to change that value no matter what.

Comment (brjeter) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: @paouelle my point is that if we change the property here, an UpdateEvent will be fired. Then this `ConfigurationInstaller` will pick up on that, see that the Felix property changed, and revert the change.

Comment (paouelle) [inline platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/FelixConfig.java]: @brjeter Good point! It also identifies a problem I commented above about making sure we only process one event at a time.
